halo2_frontend/src/dev.rs

```diff
@@ -1787,4 +1800,107 @@ mod tests {
             },])
         )
     }
+
+    #[test]
+    fn modify_proof() {
+        const K: u32 = 4;
+
+        struct EasyCircuit {}
+
+        #[derive(Clone)]
+        struct EasyCircuitConfig {
+            a: Column<Advice>,
+            b: Column<Instance>,
+            q: Selector,
+        }
+
+        impl Circuit<Fp> for EasyCircuit {
+            type Config = EasyCircuitConfig;
+            type FloorPlanner = SimpleFloorPlanner;
+            #[cfg(feature = "circuit-params")]
+            type Params = ();
+
+            fn without_witnesses(&self) -> Self {
+                Self {}
+            }
+
+            fn configure(meta: &mut ConstraintSystem<Fp>) -> Self::Config {
+                let a = meta.advice_column();
+                let b = meta.instance_column();
+                let q = meta.selector();
+
+                meta.enable_equality(a);
+                meta.enable_equality(b);
+
+                meta.create_gate("squared", |cells| {
+                    let cur = cells.query_advice(a, Rotation::cur());
+                    let next = cells.query_advice(a, Rotation::next());
+                    let q = cells.query_selector(q);
+
+                    vec![q * (next - cur.clone() * cur)]
+                });
+
+                EasyCircuitConfig { a, b, q }
+            }
+
+            fn synthesize(
+                &self,
+                config: Self::Config,
+                mut layouter: impl Layouter<Fp>,
+            ) -> Result<(), Error> {
+                let out = layouter.assign_region(
+                    || "main region",
+                    |mut region| {
+                        config.q.enable(&mut region, 0)?;
+                        config.q.enable(&mut region, 1)?;
+
+                        region.assign_advice(|| "a0", config.a, 0, || Value::known(Fp::from(3)))?;
+                        region.assign_advice(|| "a1", config.a, 1, || Value::known(Fp::from(9)))?;
+                        region.assign_advice(|| "a2", config.a, 2, || Value::known(Fp::from(10)))
```

Comment:
I would suggest having a circuit that assigns a correct witness, make sure the mock prover can verify it, and then start manipulating the advices / instances to check that it now gives errors when verifying with the mock prover.
This flow better shows how a negative test would go: start from a correct assignment that passes, make a hypothesis about a witness that should not pass, and make sure it doesn't.